src/renderer/render.rs

```diff
@@ -104,100 +104,74 @@ pub fn render_scene(
-            let b = b.transform.translation().z;
-            a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut shadow_items);
-
-        // Ground items: y ordered
-        let mut middle_items: Vec<(_, _)> = render_vectors
-            .iter()
-            .filter(|(_, v)| v.layer == Layer::Ground)
-            .collect();
-        middle_items.sort_by(|(_, a), (_, b)| {
-            let a = a.transform.translation().y;
-            let b = b.transform.translation().y;
-            b.partial_cmp(&a).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut middle_items);
-
-        // Foreground items: z ordered
-        let mut fg_items: Vec<(_, _)> = render_vectors
-            .iter()
-            .filter(|(_, v)| v.layer == Layer::Foreground)
-            .collect();
-        fg_items.sort_by(|(_, a), (_, b)| {
-            let a = a.transform.translation().z;
-            let b = b.transform.translation().z;
-            a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut fg_items);
-
-        // UI items
-        let mut ui_items: Vec<(_, _)> = render_vectors
-            .iter()
-            .filter(|(_, v)| v.layer == Layer::UI)
-            .collect();
-        ui_items.sort_by(|(_, a), (_, b)| {
-            let a = a.transform.translation().z;
-            let b = b.transform.translation().z;
-            a.partial_cmp(&b).unwrap_or(std::cmp::Ordering::Equal)
-        });
-        vector_render_queue.append(&mut ui_items);
+        enum RenderItem<'a> {
+            Vector(&'a ExtractedRenderVector),
+            Text(&'a ExtractedRenderText),
+        }
+        let mut render_queue: Vec<(f32, RenderMode, (&PreparedAffine, RenderItem))> =
+            render_vectors
+                .iter()
+                .map(|(a, b)| {
+                    (
+                        b.transform.translation().z,
+                        b.render_mode,
+                        (a, RenderItem::Vector(b)),
+                    )
+                })
+                .collect();
+        render_queue.extend(query_render_texts.iter().map(|(a, b)| {
+            (
+                b.transform.translation().z,
+                b.render_mode,
+                (a, RenderItem::Text(b)),
+            )
+        }));
+
+        // Sort by render mode with screen space on top, then by z-index
+        render_queue.sort_by(
+            |(a_z_index, a_render_mode, _), (b_z_index, b_render_mode, _)| {
+                let z_index = a_z_index.partial_cmp(b_z_index).unwrap();
+                let render_mode = if a_render_mode == b_render_mode {
+                    std::cmp::Ordering::Equal
+                } else if *a_render_mode == RenderMode::ScreenSpace {
+                    std::cmp::Ordering::Greater
+                } else {
+                    std::cmp::Ordering::Less
+                };
```

Comment:
I think it makes more sense to implement `PartialEq`, and `Eq` for `RenderMode`, which would allow us to just say `a_render_mode.cmp(b_render_mode)`

Reply:
Added Eq, PartialOrd, and Ord, and made the ordinal of the enums explicit to prevent accidental reordering